packages/neuron-wallet/src/utils/mnemonic/index.ts

```diff
@@ -0,0 +1,122 @@
+import crypto from 'crypto'
+import wordList from './wordList'
+
+const RADIX = 2048
+const PBKDF2_ROUNDS = 2048
+const KEY_LEN = 64
+const MIN_ENTROPY_SIZE = 16
+const MAX_ENTROPY_SIZE = 32
+
+const INVALID_MNEMONIC = `Invalid mnemonic`
+const INVALID_CHECKSUM = `Invalid checksum`
+const ENTROPY_NOT_DIVISIBLE = `Entropy should be divisable by 4`
+const ENTROPY_TOO_LONG = `Entropy should be shorter than ${MAX_ENTROPY_SIZE + 1}`
+const ENTROPY_TOO_SHORT = `Entropy should be longer than ${MIN_ENTROPY_SIZE - 1}`
+
+if (wordList.length !== RADIX)
+  throw new Error(`Word list should have ${RADIX} words, but ${wordList.length} received in fact`)
+
+const bytesToBinary = (bytes: Buffer): string => {
+  return bytes.reduce((binary, byte) => {
+    return binary + byte.toString(2).padStart(8, '0')
+  }, '')
+}
+
+const deriveChecksumBits = (entropyBuffer: Buffer): string => {
+  const ENT = entropyBuffer.length * 8
+  const CS = ENT / 32
+  const hash = crypto
+    .createHash('sha256')
+    .update(entropyBuffer)
+    .digest()
+  return bytesToBinary(hash).slice(0, CS)
+}
+
+const salt = (password = '') => {
+  return `mnemonic${password}`
+}
+
+export const mnemonicToSeedSync = (mnemonic: string = '', password: string = '') => {
+  const mnemonicBuffer = Buffer.from(mnemonic.normalize('NFKD'), 'utf8')
+  const saltBuffer = Buffer.from(salt(password.normalize('NFKD')), 'utf8')
+  return crypto.pbkdf2Sync(mnemonicBuffer, saltBuffer, PBKDF2_ROUNDS, KEY_LEN, 'sha512')
+}
+
+export function mnemonicToSeed(mnemonic: string = '', password: string = ''): Promise<Buffer> {
+  return new Promise((resolve, reject) => {
+    try {
+      const mnemonicBuffer = Buffer.from(mnemonic.normalize('NFKD'), 'utf8')
+      const saltBuffer = Buffer.from(salt(password.normalize('NFKD')), 'utf8')
+      crypto.pbkdf2(mnemonicBuffer, saltBuffer, PBKDF2_ROUNDS, KEY_LEN, 'sha512', (err, data) => {
+        if (err) reject(err)
+        resolve(data)
+      })
+    } catch (error) {
+      reject(error)
+    }
+  })
+}
+
+export function mnemonicToEntropy(mnemonic: string = '') {
+  const words = mnemonic.normalize('NFKD').split(' ')
+  if (words.length % 3 !== 0) throw new Error(INVALID_MNEMONIC)
```

Comment:
Practically beside `% 3 === 0`, the length could only be one of `12` - `24`. Would it be necessary to check that as well to catch length validation error early?

Reply:
> Practically beside `% 3 === 0`, the length could only be one of `12` - `24`. Would it be necessary to check that as well to catch length validation error early?

updated in https://github.com/nervosnetwork/neuron/pull/439/commits/7cddc8e5b72bb18dd0dcecaae8b301fbc84bb49c